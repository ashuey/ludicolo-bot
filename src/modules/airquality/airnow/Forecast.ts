import { Category } from "@/modules/airquality/airnow/Category";

export interface Forecast {
    DateIssue: string;
    DateForecast: string;
    ReportingArea: string;
    StateCode: string;
    Latitude: number;
    Longitude: number;
    ParameterName: string;
    AQI: number;
    Category: Category;
    ActionDay: boolean;
    Discussion: string;
}
