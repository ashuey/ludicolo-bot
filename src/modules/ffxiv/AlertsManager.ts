import { EmbedBuilder } from "discord.js";
import { PriceSnipeEvent, TataruClient } from "@/modules/ffxiv/lib/tataru/TataruClient";
import { Application } from "@/common/Application";
import { logger } from "@/logger";

// TODO: Don't hardcode
const alertChannels = ['1277359699703894046'];
const alertChannelsProd = ['1277776716772937799'];
const alertChannelsGilgamesh = ['1277709841661689990'];
const alertChannelsGilgameshProd = ['1277776772364238938'];

export class AlertsManager {
    static GilFormatter = new Intl.NumberFormat('en-US', {maximumFractionDigits: 0});

    static QuantityFormatter = new Intl.NumberFormat('en-US', {maximumFractionDigits: 0});

    static VelocityFormatter = new Intl.NumberFormat('en-US', {maximumFractionDigits: 1});

    protected readonly app: Application;

    protected readonly tataru: TataruClient;

    protected active = false;

    constructor(app: Application) {
        this.app = app;
        this.tataru = new TataruClient(app.config.tataruApiKey);
        this.tataru.onAlert((e) => this.handle(e));
    }

    connect() {
        this.active = true;
        this.tataru.connect();
    }

    refresh() {
        if (this.active) {
            this.tataru.refresh();
        }
    }

    protected async handle(e: PriceSnipeEvent) {
        const channelsForAlert = this.app.isProduction ?
            // TODO: Don't hardcode Gilgamesh check
            (e.world === "Gilgamesh" ? [...alertChannelsProd, ...alertChannelsGilgameshProd] : [...alertChannelsProd])
            : (e.world === "Gilgamesh" ? [...alertChannels, ...alertChannelsGilgamesh] : [...alertChannels]);
        for (const channelId of channelsForAlert) {
            const channel = await this.app.discord.channels.fetch(channelId);

            if (!channel) {
                logger.warn(`Channel for XIV Alerts was not found: ${channelId}`);
                continue;
            }

            if (!channel.isTextBased()) {
                logger.warn(`Channel for XIV Alerts is not text-based: ${channel.name}`);
                continue;
            }

            await channel.send({
                embeds: [this.buildEmbed(e)],
            })
        }
    }

    protected buildEmbed(e: PriceSnipeEvent): EmbedBuilder {
        return new EmbedBuilder()
            .setTitle(`Price Snipe Alert: ${e.itemName} (${e.hq ? 'HQ' : 'NQ'})`)
            .setDescription(`ðŸŒŽ ${e.world}`)
            .setColor(e.world === "Gilgamesh" ? "#2ecc71" : "#f1c40f")
            .addFields(
                {
                    name: "Previous Price",
                    value: `${AlertsManager.GilFormatter.format(e.previousPrice)} gil`,
                    inline: true
                },
                {
                    name: "New Price",
                    value: `${AlertsManager.GilFormatter.format(e.price)} gil`,
                    inline: true
                },
                {
                    name: "Price Delta",
                    value: `${AlertsManager.GilFormatter.format(e.previousPrice - e.price)} gil`,
                    inline: false
                },
                {
                    name: "Quantity",
                    value: AlertsManager.QuantityFormatter.format(e.quantity),
                    inline: false
                },
                {
                    name: "Daily Sales Velocity",
                    value: `${AlertsManager.VelocityFormatter.format(e.velocity)} sales per day`,
                    inline: false
                },
            );
    }
}
