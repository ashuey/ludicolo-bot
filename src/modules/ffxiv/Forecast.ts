import EorzeaWeather from "eorzea-weather";
import {ForecastEntry} from "@/modules/ffxiv/weather/ForecastEntry";

const ET_ONE_HOUR = 175 * 1000;
const ET_EIGHT_HOURS = 8 * ET_ONE_HOUR;

export class Forecast {
    protected readonly eorzeaWeather: EorzeaWeather;

    protected forecast: ForecastEntry[] = [];

    constructor(zone: string) {
        this.eorzeaWeather = new EorzeaWeather(zone);
        this.refresh();
    }

    refresh() {
        const now = new Date();
        const startTime = this.getStartTime(now.getTime());
        this.forecast = [];
        for (let time = startTime - (ET_EIGHT_HOURS * 6); time < (startTime + (ET_EIGHT_HOURS * 6)); time += ET_EIGHT_HOURS) {
            const startedAt = new Date(time);
            this.forecast.push({
                name: this.eorzeaWeather.getWeather(new Date(time)),
                startedAt,
            });
        }
    }

    findNext(weather: string): ForecastEntry | undefined {
        const now = new Date();

        return this.forecast.find((entry, idx) =>
            entry.name.toLowerCase() === weather.toLowerCase()
            && entry.startedAt > now
            && entry.name !== this.forecast[idx - 1]?.name
        );
    }

    findPrevious(reference: ForecastEntry, maxLookback: number = 18): ForecastEntry | undefined {
        let offset = 0;
        let thisWeather = '';
        let lastMatch: ForecastEntry | undefined = undefined;
        let outOfCurrentWindow = false;

        do {
            offset += 1;
            const startedAt = new Date(reference.startedAt.getTime() - (offset * ET_EIGHT_HOURS));
            thisWeather = this.eorzeaWeather.getWeather(startedAt);

            if (outOfCurrentWindow && thisWeather === reference.name) {
                lastMatch = {
                    name: thisWeather,
                    startedAt,
                }
            }

            if (!outOfCurrentWindow && thisWeather !== reference.name) {
                outOfCurrentWindow = true;
            }
        } while (offset < maxLookback && (!outOfCurrentWindow || lastMatch === undefined || thisWeather === reference.name));

        return lastMatch;
    }

    current(): string {
        return this.eorzeaWeather.getWeather(new Date());
    }

    protected getStartTime = (msec: number): number => {
        const bell = (msec / ET_ONE_HOUR) % 24;
        return msec - Math.round(ET_ONE_HOUR * bell);
    };
}


