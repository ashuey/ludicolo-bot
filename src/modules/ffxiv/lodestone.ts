import { JSDOM } from "jsdom";
import {logger} from "@/logger";

const userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36";
const userAgentMobile = "Mozilla/5.0 (iPhone; CPU OS 10_15_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/14E304 Safari/605.1.15";
const achievementIdRegex = /lodestone\/character\/\d*\/achievement\/detail\/(\d*)\//;

export enum LODESTONE_RESULT {
    SUCCESS,
    ERROR,
    PRIVATE,
    NOT_FOUND,
}

export async function getAchievements(characterId: string): Promise<[LODESTONE_RESULT, string[]]> {
    try {
        const response = await fetch(`https://na.finalfantasyxiv.com/lodestone/character/${characterId}/achievement/`, {
            headers: {
                'user-agent': userAgent
            }
        });
        const body = await response.text();
        if (response.status === 403 && body.includes('You do not have permission to view this page.')) {
            return [LODESTONE_RESULT.PRIVATE, []];
        }

        if (response.status === 404 && body.includes('The page you are searching for has either been removed, or the designated URL address is incorrect')) {
            logger.warn(`Error ${response.status} while fetching achievements for ${characterId}`);
            return [LODESTONE_RESULT.NOT_FOUND, []];
        }

        if (response.status !== 200) {
            logger.warn(`Error ${response.status} while fetching achievements for ${characterId}`);
            return [LODESTONE_RESULT.ERROR, []];
        }

        const { document } = new JSDOM(body).window;
        const nodes = document.querySelectorAll('.ldst__achievement .entry');
        const achievements = [...nodes].map((node, idx) => {
            const idNode = node.querySelector('.entry__achievement');

            if (!idNode) {
                throw new Error(`Could not find ID node in achievement (Character: ${characterId}, Index: ${idx})`);
            }

            const link = idNode.getAttribute('href');

            if (!link) {
                throw new Error(`ID node had no href attribute (Character: ${characterId}, Index: ${idx})`);
            }

            const regexResult = achievementIdRegex.exec(link);

            if (!regexResult || !regexResult[1]) {
                throw new Error(`Could not extract achievement ID from link: ${link}`);
            }

            return regexResult[1];
        });

        return [LODESTONE_RESULT.SUCCESS, achievements];
    } catch (e) {
        logger.error(e);
        return [LODESTONE_RESULT.ERROR, []];
    }
}

export async function getMounts(characterId: string): Promise<[LODESTONE_RESULT, string[]]> {
    try {
        const response = await fetch(`https://na.finalfantasyxiv.com/lodestone/character/${characterId}/mount/`, {
            headers: {
                'user-agent': userAgentMobile
            }
        });
        const body = await response.text();

        if (response.status === 404 && body.includes('The page you are searching for has either been removed, or the designated URL address is incorrect')) {
            logger.warn(`Error ${response.status} while fetching achievements for ${characterId}`);
            return [LODESTONE_RESULT.NOT_FOUND, []];
        }

        if (response.status !== 200) {
            logger.warn(`Error ${response.status} while fetching mounts for ${characterId}`);
            return [LODESTONE_RESULT.ERROR, []];
        }

        const {document} = new JSDOM(body).window;
        const nodes = document.querySelectorAll('.mount__list__item');
        const mounts = [...nodes].map((node, idx) => {
            const nameNode = node.querySelector('.mount__name');

            if (!nameNode) {
                throw new Error(`Could not find name node in mount (Character: ${characterId}, Index: ${idx})`);
            }

            const name = nameNode.textContent;

            if (!name) {
                throw new Error(`Name node had no text content (Character: ${characterId}, Index: ${idx})`);
            }

            return name;
        });

        return [LODESTONE_RESULT.SUCCESS, mounts];
    } catch (e) {
        logger.error(e);
        return [LODESTONE_RESULT.ERROR, []];
    }
}
