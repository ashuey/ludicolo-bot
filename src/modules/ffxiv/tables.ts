export const STATIC_DATA_TABLE = 'ffxiv_static_data';
