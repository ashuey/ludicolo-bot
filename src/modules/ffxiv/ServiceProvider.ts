import {AlertsManager} from "@/modules/ffxiv/alerts/AlertsManager";
import { Module } from "@/common/Module";
import { Application } from "@/common/Application";

export interface ServiceProvider extends Module {
    readonly app: Application;
    readonly alerts: AlertsManager;
}
