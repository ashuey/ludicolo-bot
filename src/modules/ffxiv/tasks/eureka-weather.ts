import {ApplicationProvider} from "@/common/ApplicationProvider";
import EorzeaWeather from "eorzea-weather";
import {Forecast} from "@/modules/ffxiv/Forecast";
import * as console from "node:console";

// TODO: This whole thing is kind of a mess

const CHANNEL_ID = "1234701925698506792"; // TODO: Don't hardcode this
//  const CHANNEL_ID = "1107395046900236388"; // DEV


let lastSentCrab = 0;
let lastSentCassie = 0;
let lastSentSkoll = 0;

export async function sendEurekaWeather(module: ApplicationProvider) {
    const channel = await module.app.discord.channels.fetch(CHANNEL_ID)
        .catch(err => {
            console.error(`Error while fetching eureka weather channel: ${err}`);
        });

    if (!channel) {
        console.warn(`Channel for Eureka weather was not found`);
        return;
    }

    if (!channel.isTextBased()) {
        console.warn(`Channel for Eureka weather is not text-based: ${channel.name}`);
        return;
    }


    const pagosForecast = new Forecast(EorzeaWeather.ZONE_EUREKA_PAGOS);
    const pyrosForecast = new Forecast(EorzeaWeather.ZONE_EUREKA_PYROS);

    const nextKingArthro = pagosForecast.findNext("fog");
    const nextCassie = pagosForecast.findNext("blizzards");
    const nextSkoll = pyrosForecast.findNext("blizzards");

    if (nextKingArthro && lastSentCrab !== nextKingArthro.startedAt.getTime()) {
        //console.log('sending ka');
        const arthroTimestamp = Math.floor(nextKingArthro.startedAt.getTime() / 1000);
        await channel.send({
            content: `[PAGOS] Next KA/Crab weather (Fog) is <t:${arthroTimestamp}:R> (<t:${arthroTimestamp}>)`,
            files: ['embeds/crabmap.png']
        }).catch(err => {
            console.error(err);
        });
        lastSentCrab = nextKingArthro.startedAt.getTime();
    }

    if (nextCassie && lastSentCassie !== nextCassie.startedAt.getTime()) {
        //console.log('sending cass');
        const cassieTimestamp = Math.floor(nextCassie.startedAt.getTime() / 1000);
        await channel.send({
            content: `[PAGOS] Next Cassie weather (Blizzards) is <t:${cassieTimestamp}:R> (<t:${cassieTimestamp}>)`,
            files: ['embeds/cassiemap.png']
        }).catch(err => {
            console.error(err);
        });
        lastSentCassie = nextCassie.startedAt.getTime();
    }

    if (nextSkoll && lastSentSkoll !== nextSkoll.startedAt.getTime()) {
        //console.log('sending skoll');
        const skollTimestamp = Math.floor(nextSkoll.startedAt.getTime() / 1000);
        await channel.send({
            content: `[PYROS] Next Skoll weather (Blizzards) is <t:${skollTimestamp}:R> (<t:${skollTimestamp}>)`,
            files: ['embeds/skollmap.png']
        }).catch(err => {
            console.error(err);
        });
        lastSentSkoll = nextSkoll.startedAt.getTime();
    }
}
