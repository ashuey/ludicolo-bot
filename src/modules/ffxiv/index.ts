import { Knex } from "knex";
import { Module } from "@/common/Module";
import { Application } from "@/common/Application";
import { ScheduledTask } from "@/common/ScheduledTask";
import { CrabHitHandler, CrabSimCommand } from "@/modules/ffxiv/cmd/crabsim";
import { ComponentHandler } from "@/common/ComponentHandler";
import { sendEurekaWeather } from "@/modules/ffxiv/tasks/eureka-weather";
import { MsqCommand } from "@/modules/ffxiv/cmd/msq";
import { DynamisCommand } from "@/modules/ffxiv/cmd/dynamis";
import { ThunderGodCommand } from "@/modules/ffxiv/cmd/thundergod";
import { XIVCommand } from "@/modules/ffxiv/cmd/xiv";
import { create_static_data_table } from "@/modules/ffxiv/migrations/create_static_data_table";
import { StaticDataManager } from "@/modules/ffxiv/lib/staticdata";
import { AlertsManager } from "@/modules/ffxiv/AlertsManager";

export class FFXIVModule implements Module {
    readonly name = 'ffxiv';

    readonly commands = [
        new XIVCommand(),
        new CrabSimCommand(),
        new MsqCommand(),
        new DynamisCommand(),
        new ThunderGodCommand(),
    ]

    readonly scheduledTasks: [string, ScheduledTask][] = [
        ['*/6 * * * * *', () => sendEurekaWeather(this)],
        ['* * * * * *', () => this.alerts.heartbeat()],
    ];

    readonly componentHandlers: [string, ComponentHandler][] = [
        ['hit_crab', new CrabHitHandler()],
    ];

    readonly migrations: [string, Knex.Migration['up']][] = [
        ['create_static_data_table', create_static_data_table]
    ]

    public readonly app: Application;

    public readonly staticData: StaticDataManager;

    public readonly alerts: AlertsManager;

    constructor(app: Application) {
        this.app = app;
        this.staticData = new StaticDataManager(app.db);
        this.alerts = new AlertsManager(app);
    }

    async bootstrap() {
        this.alerts.connect();
    }
}
