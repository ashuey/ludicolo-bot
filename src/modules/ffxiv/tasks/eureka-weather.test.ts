import {forecastIsAlertable, TWENTY_MINUTES} from "@/modules/ffxiv/tasks/eureka-weather";
import {ForecastEntry} from "@/modules/ffxiv/Forecast";

describe('eureka-weather', () => {
    it('twenty minutes is twenty minutes', () => {
        const startTime = 1714791600000; // Sat May 04 2024 03:00:00 GMT+0000
        const expectedTime = 1714792800000; // Sat May 04 2024 03:20:00 GMT+0000

        expect(startTime + TWENTY_MINUTES).toBe(expectedTime);
    });

    describe('forecastIsAlertable', () => {
        beforeAll(() => {
            jest.useFakeTimers();
        });

        afterAll(() => {
            jest.useRealTimers();
        });

        it('alerts for events within twenty minutes', () => {
            jest.setSystemTime(1714782630000) // Sat May 04 2024 00:30:30 GMT+0000

            const forecast: ForecastEntry = {
                name: 'Snow',
                startedAt: new Date(1714782640000), // Sat May 04 2024 00:30:40 GMT+0000
            }

            expect(forecastIsAlertable(forecast, 0)).toBe(true);
        });

        it('does not alert for events more than twenty minutes in the future', () => {
            jest.setSystemTime(1714742400000) // Fri May 03 2024 13:20:00 GMT+0000

            const forecast: ForecastEntry = {
                name: 'Snow',
                startedAt: new Date(1714744560000), // Fri May 03 2024 13:56:00 GMT+0000
            }

            expect(forecastIsAlertable(forecast, 0)).toBe(false);
        });

        it('does not alert for events that have already been alerted', () => {
            jest.setSystemTime(1714749120000) // Fri May 03 2024 15:12:00 GMT+0000

            const forecast: ForecastEntry = {
                name: 'Snow',
                startedAt: new Date(1714749600000), // Fri May 03 2024 15:20:00 GMT+0000
            }

            expect(forecastIsAlertable(forecast, 1714749600000)).toBe(false);
        });
    })
});
