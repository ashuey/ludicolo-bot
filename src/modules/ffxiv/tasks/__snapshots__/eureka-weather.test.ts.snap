// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eureka-weather refactor snapshots Returns the correct resonse when time is 1716460123534 1`] = `[]`;

exports[`eureka-weather refactor snapshots Returns the correct resonse when time is 1716483013917 1`] = `
[
  [
    {
      "embeds": [
        {
          "color": 7447510,
          "description": "🕛  <t:1716484000:R> (<t:1716484000:t>)

🗺  Eureka Pyros  —  Northpoint",
          "title": "🐺 🐺 🐺  Skoll Spawn Window 🐺 🐺 🐺",
        },
      ],
    },
  ],
]
`;

exports[`eureka-weather refactor snapshots Returns the correct resonse when time is 1716487212135 1`] = `
[
  [
    {
      "embeds": [
        {
          "color": 11396636,
          "description": "🕛  <t:1716488200:R> (<t:1716488200:t>)

✅ Guaranteed Spawn. Last window ended 233m earlier

🗺  Eureka Pagos  —  Gravitational Studies",
          "title": "🪴 🪴 🪴  Copycat Cassie Spawn Window 🪴 🪴 🪴",
        },
      ],
    },
  ],
]
`;

exports[`eureka-weather refactor snapshots Returns the correct resonse when time is 1716488596352 1`] = `
[
  [
    {
      "embeds": [
        {
          "color": 16741376,
          "description": "🕛  <t:1716489600:R> (<t:1716489600:t>)

❕ Last spawn 116m earlier

🗺  Eureka Pagos  —  Geothermal Studies",
          "title": "🦀 🦀 🦀  King Arthro Spawn Window 🦀 🦀 🦀",
        },
      ],
    },
  ],
]
`;
